### PR TITLE
*: update txn commit

### PR DESCRIPTION
**Pull request by disksing**
@morefreeze @ngaut @shenli

### CODE REVIEW

File: txn_committer.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+				Key: k,
+			}
+		}
+		keys = append(keys, k)
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Transactions with no Put/Del, only Locks are readonly.
```
Comment (morefreeze) [inline txn_committer.go]: with no -> without

File: txn.go [hunk 2]
```diff
@@ -104,253 +102,23 @@ func (txn *tikvTxn) DelOption(opt kv.Option) {
...
-				// Primary key must be in first doCommit.
-				if reqIdx == 0 && startIdx == 0 {
-					return errors.Trace(err)
-				}
-				// It is ok even if secondary key commit failed.
-				log.Warnf("Commit error(not primary): %s", err)
-			}
-		}
-	}
 	return nil
```
Comment (morefreeze) [inline txn.go]: `return errors.Trace(err)` ?

Comment (disksing) [inline txn.go]: @morefreeze Already got a `return errors.Trace(err)` above.

File: txn_committer.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+				Key: k,
+			}
+		}
+		keys = append(keys, k)
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Transactions without Put/Del, only Locks are readonly.
```
Comment (shenli) [inline txn_committer.go]: Select for update?

File: txn_committer.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		resp, err := c.store.SendKVReq(req, region)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if regionErr := resp.GetRegionError(); regionErr != nil {
+			// re-split keys and prewrite again.
+			err = c.iterKeysByRegion(keys, c.prewriteSingleRegion)
```
Comment (shenli) [inline txn_committer.go]: Is it possible that you meet infinite recursive?

**Comment by shenli** [comment]
Rest LGTM

**Comment by morefreeze** [comment]
I think you can do some cleaning like `txn.primary` in next PR.
LGTM

**Comment by shenli** [comment]
LGTM

File: txn_committer.go (hunk 3 above)
Comment (disksing) [inline txn_committer.go]: Yes. Maybe executed a `Select for update` statement but no further write.

**Pull request merged**